tensorflow-framework/src/main/java/org/tensorflow/framework/optimizers/Optimizer.java

```diff
@@ -323,13 +364,15 @@ public float getLearningRate() {
    *     Operand has been set.
    */
   public Map<Operand<? extends TType>, Tensor<? extends TType>> getFeedMap() {
-    return this.feedMap;
+    return feedMap;
   }
 
+  /** {@inheritDoc}  */
   public void close() {
     // close the learningRate Tensor if it exists.
-    if (this.feedMap != null) {
-      this.feedMap.get(this.learningRatePlaceholder).close();
+    if (learningRateTensor != null) {
+      learningRateTensor.close();
+      learningRateTensor = null;
```

Comment:
This still leaves a valid reference in the `feedMap` though the tensor it refers to is closed.

Reply:
OK, but I am starting to question in my mind, whether we should use `PlaceHolder` within the `Optimizer` itself, but rather store the `learningRate` as an `Operand`. This allows the user of the `Optimizer` to define the learning rate as needed, For example,  a `Constant`, `Placeholder`, `Variable`, or an `Operand` such as what may be defined in a scheduler.  If we did this, it would be up to the `Optimizer` user, to make the change to the learning rate, via one of the mechanisms just outlined as an `Operand`.  Also, if we did this general approach, we could get rid of `feedMap`.  

In reality, `feedMap` is just a pair tying a `Placeholder` to a `Tensor`.  When you use `feed` in `Session.Runner`, each member of the pair is passed as two distinct parameters.

Reply:
So I think that having it be an operand is probably reasonable, as then the learning rate scheduler can own it, but I think the current approach is the most flexible (as it means users don't have to do the scheduler calculation in TensorFlow if they don't want to). If it's an operand then we'd need to implement a variety of learning rate schedulers, and I'm not sure if that's in the set of things you've already done or not.

Reply:
OK, I'll leave it as is with the Placeholder. 

I do have scheduler classes that build their own graphs for calculating the new learning rate based on the current training step. This is usually done be adding a `Variable` into that generated graph, and updating it for each training step. However,  I was going to add those in another PR later. For that feature,  I would need to change `Optimizer`  to add a `setLearningRateOperand` method. I already have a `getLearningRateOperand` method that currently returns the Placeholder for the `Optimizer` subclasses to use when building their graphs. My plan was to have `getLearningRateOperand`  return either the internal `Placeholder` or the new `Operand` based on how the `Optimizer` was constructed. Should I add that method now or wait?

Reply:
May as well add it now if you've already got it written. It'll change the logic in here anyway and we may as well see all of it in one go.

Reply:
You want me to add the schedules?. They are currently in package `org.tensorflow.framework.optimizers.schedules` and include `ExponentialDecay`, `InverseTimeDecay`, `PiecewiseConstantDecay` and `PolynomialDecay`.

Reply:
Erm, probably not, lets try to keep the PRs small-ish.

Reply:
I have added `Operand<TFloat32> learningRateOperand` to `Optimizer` and kept `Placeholder<TFloat32> learningRatePlaceholder`. I also added a constructor to set the `Operand` and kept the one that uses a  `float` for learning rate. 
`getLearningRateOperand` now returns the `learningRatePlaceholder ` if it exists, otherwise it returns the `learningRateOperand`.